File: src/shared/models/transaction_category_balance.py
# ============================================================================
# Source: COBOL copybook CVTRA01Y.cpy — TRAN-CAT-BAL-RECORD (RECLN 50)
# ============================================================================
# Mainframe-to-Cloud migration: VSAM KSDS → Aurora PostgreSQL
#
# Replaces the mainframe TCATBALF VSAM KSDS cluster (see
# ``app/jcl/TCATBALF.jcl``) with a relational PostgreSQL table persisting
# per-account, per-(transaction-type, transaction-category) balance rows.
# The single monetary column ``balance`` uses PostgreSQL NUMERIC(15, 2) to
# preserve the exact COBOL PIC S9(09)V99 decimal semantics — no floating-
# point arithmetic is permitted on this field (see AAP §0.7.2 "Financial
# Precision").
# ============================================================================
# Copyright Amazon.com, Inc. or its affiliates.
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""SQLAlchemy 2.x ORM model for the ``transaction_category_balance`` table.

Converts the COBOL copybook ``app/cpy/CVTRA01Y.cpy`` (record layout
``TRAN-CAT-BAL-RECORD``, 50-byte fixed-width record) to a SQLAlchemy 2.x
declarative ORM model representing a per-account running balance keyed
by ``(account, transaction-type, transaction-category)`` in the CardDemo
Aurora PostgreSQL database.

The transaction-category-balance table is the runtime accrual ledger for
credit-card account balances broken down by transaction-type and
transaction-category — a normalized replacement for the mainframe VSAM
KSDS cluster ``TCATBALF``. It is consumed by:

* **Batch posting** (``CBTRN02C`` → ``src/batch/jobs/posttran_job.py``)
  to apply successful transaction postings by adding (or subtracting,
  per the transaction type) the transaction amount to the relevant
  category balance during Stage 1 of the nightly pipeline.
* **Interest calculation** (``CBACT04C`` →
  ``src/batch/jobs/intcalc_job.py``) to iterate all category balances
  for every account and apply the canonical interest formula
  ``(TRAN-CAT-BAL × DIS-INT-RATE) / 1200`` from the disclosure-group
  rate table — Stage 2 of the nightly pipeline.
* **Statement generation** (``CBSTM03A`` →
  ``src/batch/jobs/creastmt_job.py``) to emit per-category balance
  subtotals into the generated statements — Stage 4a of the nightly
  pipeline.

COBOL to Python Field Mapping
-----------------------------
==================  =================  ==================  =========================
COBOL Field         COBOL Type         Python Column       SQLAlchemy Type
==================  =================  ==================  =========================
TRANCAT-ACCT-ID     ``PIC 9(11)``      ``acct_id``         ``String(11)`` — PK #1
TRANCAT-TYPE-CD     ``PIC X(02)``      ``type_cd``         ``String(2)``  — PK #2
TRANCAT-CD          ``PIC 9(04)``      ``cat_cd``          ``String(4)``  — PK #3
TRAN-CAT-BAL        ``PIC S9(09)V99``  ``balance``         ``Numeric(15, 2)`` †
FILLER              ``PIC X(22)``      — (not mapped)      — (COBOL padding only)
==================  =================  ==================  =========================

† **Monetary field.** ``Numeric(15, 2)`` maps to PostgreSQL
  ``NUMERIC(15, 2)``: 15 total digits with exactly 2 decimal places,
  which accommodates the full signed range of COBOL ``PIC S9(09)V99``
  (11 digits total = 9 integer + 2 decimal) with comfortable headroom
  for any downstream widening of the monetary precision. All
  arithmetic on this field uses :class:`decimal.Decimal` with banker's
  rounding (``ROUND_HALF_EVEN``) to preserve the COBOL ``ROUNDED``
  clause semantics. Floating-point arithmetic is **never** permitted
  on this column (AAP §0.7.2 "Financial Precision").

Total RECLN: 11 + 2 + 4 + 11 + 22 = 50 bytes — matches the VSAM
cluster definition in ``app/jcl/TCATBALF.jcl`` (``RECSZ(50 50)``) and
the original COBOL copybook record length.

Composite Primary Key — ``TRAN-CAT-KEY``
----------------------------------------
The 3-part composite primary key ``(acct_id, type_cd, cat_cd)``
directly mirrors the COBOL group-level key ``TRAN-CAT-KEY``
(lines 5–8 of ``CVTRA01Y.cpy``), which is the full key used for
``READ``/``WRITE``/``REWRITE``/``DELETE`` operations against the
originating VSAM KSDS dataset ``TCATBALF``. Preserving the three
fields as composite PK components (rather than introducing a surrogate
integer ID) keeps the relational row addressable by the same logical
key as its VSAM ancestor — a prerequisite for behaviour-preserving
batch translation (see AAP §0.7.1 "Refactoring-Specific Rules" —
preserve existing business logic without modification).

The key ordering ``(acct_id, type_cd, cat_cd)`` is chosen deliberately:
the high-cardinality ``acct_id`` leads the key so that per-account
range scans (performed by both the posting and interest-calculation
jobs to fetch an account's full balance breakdown in a single
indexed read) are served by the primary key's B-tree without
requiring a secondary index.

Financial Precision — ``Numeric(15, 2)``
----------------------------------------
The ``balance`` column uses PostgreSQL ``NUMERIC(15, 2)`` — a
fixed-precision decimal with 15 total digits and 2 fractional digits.
This exactly represents (and provides headroom above) the COBOL
declaration ``PIC S9(09)V99`` (9 integer digits + 2 fractional
digits, implicit decimal point, signed). SQLAlchemy returns this
column as a ``decimal.Decimal`` value, never a Python ``float``.
Floating-point arithmetic is **forbidden** on any financial value in
the CardDemo application per AAP §0.7.2 "Financial Precision".

The column is declared ``nullable=False`` with
``default=Decimal("0.00")``. A missing balance would break the
downstream posting and interest-accrual formulas; defaulting to zero
yields a safe, arithmetically-neutral row when a brand-new
(account, type, category) combination is first written.

Interest-Calculation Consumer
-----------------------------
The stored balance is consumed by the canonical interest formula
from ``CBACT04C.cbl`` — which, per AAP §0.7.1 "Refactoring-Specific
Rules", must not be algebraically simplified::

    monthly_interest = (tran_cat_bal * dis_int_rate) / 1200

where ``tran_cat_bal`` is this column (``balance``) and ``dis_int_rate``
is the matching rate from ``disclosure_group.int_rate``. The literal
``1200`` — representing 12 months × 100 percent — is preserved as
specified in the AAP.

Design Notes
------------
* Uses SQLAlchemy 2.x :func:`~sqlalchemy.orm.mapped_column` style with
  typed :class:`~sqlalchemy.orm.Mapped` annotations (NOT the legacy
  ``Column()`` constructor).
* ``Base`` is imported from the package ``__init__.py``
  (``src.shared.models``) so that this entity registers with the
  shared :class:`~sqlalchemy.MetaData` alongside the other CardDemo
  models — enabling unified schema management, Alembic migrations,
  and ``Base.metadata.create_all()`` bootstrap.
* No ``FILLER`` column is mapped — the trailing 22 bytes of COBOL
  padding have no relational counterpart. In PostgreSQL, column
  widths are explicit and trailing padding has no storage or
  semantic meaning.
* ``TRANCAT-ACCT-ID`` is declared in COBOL as ``PIC 9(11)`` (unsigned
  numeric, 11 digits) but is mapped to ``String(11)`` in the
  relational model to preserve leading-zero formatting (e.g.,
  ``'00000000001'`` vs ``'1'``). This keeps the account-id
  semantically a label rather than a numeric quantity — consistent
  with ``Account.acct_id`` in the sibling model — and avoids lossy
  round-trips when joining against the ``account`` table.
* ``TRANCAT-CD`` is declared in COBOL as ``PIC 9(04)`` (unsigned
  numeric, 4 digits) but is mapped to ``String(4)`` in the
  relational model to preserve leading-zero formatting (e.g.,
  ``'0001'`` vs ``'1'``). This keeps the code semantically a label
  rather than a numeric quantity — consistent with
  ``DisclosureGroup.tran_cat_cd`` and ``TransactionCategory.cat_cd``
  in sibling models — and avoids lossy round-trips when joining
  against those tables.
* Column names (``acct_id``, ``type_cd``, ``cat_cd``, ``balance``)
  are specified in AAP §0.5.1 "File-by-File Transformation Plan"
  and intentionally drop the ``tran_`` prefix of the COBOL field
  names because within the ``transaction_category_balance`` table
  the ``tran_`` prefix is implicit in the table name itself.
* No relationships (``relationship()``) are declared on this entity
  to keep it free of back-references and avoid circular-import risk
  while the model package is being assembled file-by-file.
  Consumers that need to join on the composite key do so explicitly
  via the PK columns against ``account.acct_id``,
  ``transaction_type.tran_type``, and the composite
  ``(type_cd, cat_cd)`` of ``transaction_category``.
* Python 3.11+ only (aligned with AWS Glue 5.1 runtime and FastAPI
  deployment baseline).

See Also
--------
AAP §0.2.2 — Batch Program Classification (POSTTRAN, INTCALC,
CREASTMT consumers of this table).
AAP §0.5.1 — File-by-File Transformation Plan
(``transaction_category_balance.py`` entry;
``posttran_job.py``/``intcalc_job.py``/``creastmt_job.py``
consumers).
AAP §0.7.1 — Refactoring-Specific Rules (preserve
``(balance × rate) / 1200`` formula unmodified; preserve the
3-part composite primary key matching ``TRAN-CAT-KEY``).
AAP §0.7.2 — Financial Precision (``Decimal`` semantics on the
``balance`` column; no floating-point arithmetic permitted).
``app/cpy/CVTRA01Y.cpy`` — Original COBOL record layout (source artifact).
``app/jcl/TCATBALF.jcl`` — Original VSAM cluster definition and
provisioning job.
``app/data/ASCII/tcatbal.txt`` — Original seed-data fixture
(50 category-balance rows).
``app/cbl/CBTRN02C.cbl`` — Original COBOL transaction-posting program
(writer of this balance table).
``app/cbl/CBACT04C.cbl`` — Original COBOL interest-calculation program
(reader of this balance table).
``app/cbl/CBSTM03A.CBL`` — Original COBOL statement-generation program
(reader of this balance table for per-category subtotals).
``src.shared.models.disclosure_group.DisclosureGroup`` — Rate-lookup
table that pairs with this balance row via
``(tran_type_cd, tran_cat_cd)`` during interest calculation.
``src.shared.models.transaction_category.TransactionCategory`` —
Reference-data table defining the valid ``(type_cd, cat_cd)`` pairs
that can appear in this balance table.
``src.shared.models.account.Account`` — Parent account record for
the ``acct_id`` composite-key component.
"""

from decimal import Decimal

from sqlalchemy import Numeric, String
from sqlalchemy.orm import Mapped, mapped_column

from src.shared.models import Base


class TransactionCategoryBalance(Base):
    """ORM entity for ``transaction_category_balance`` (from COBOL ``TRAN-CAT-BAL-RECORD``).

    Represents a single row in the CardDemo Aurora PostgreSQL
    ``transaction_category_balance`` table, which replaces the
    mainframe VSAM KSDS ``TCATBALF`` dataset. Each row records the
    running balance for a single
    ``(account, transaction-type, transaction-category)`` tuple and
    is read/written by:

    * **Stage 1 — POSTTRAN** (``CBTRN02C`` →
      ``posttran_job.py``). For every valid posting, the
      transaction amount is added (or subtracted per type) to the
      corresponding ``(acct_id, type_cd, cat_cd)`` row's
      ``balance``, creating the row with a zero default if it does
      not already exist.
    * **Stage 2 — INTCALC** (``CBACT04C`` → ``intcalc_job.py``).
      For each row, the monthly interest is computed as
      ``(balance × disclosure_group.int_rate) / 1200`` and applied
      to the account's current balance. The literal ``1200`` — 12
      months × 100 percent — is preserved verbatim in the batch
      code per AAP §0.7.1.
    * **Stage 4a — CREASTMT** (``CBSTM03A`` →
      ``creastmt_job.py``). Per-category balance subtotals are
      rolled up per account and rendered into the generated
      statement (text and HTML variants).

    Attributes
    ----------
    acct_id : str
        **Primary key part 1 of 3.** 11-character account identifier
        (from COBOL ``TRANCAT-ACCT-ID``, ``PIC 9(11)``). Stored as
        ``String(11)`` with leading zeros preserved (e.g.,
        ``'00000000001'``) to match the fixed-width label semantics
        of ``Account.acct_id`` for clean composite-key joins.
    type_cd : str
        **Primary key part 2 of 3.** 2-character transaction-type
        code (from COBOL ``TRANCAT-TYPE-CD``, ``PIC X(02)``).
        Logically references the ``transaction_type.tran_type``
        reference-data table (e.g., ``'01'`` = Purchase,
        ``'02'`` = Payment, ``'03'`` = Credit, ``'04'`` = Debit,
        ``'05'`` = Refund, ``'06'`` = Adjustment, ``'07'`` = Fee).
    cat_cd : str
        **Primary key part 3 of 3.** 4-character transaction-category
        code (from COBOL ``TRANCAT-CD``, ``PIC 9(04)``). Stored as a
        fixed-width string with leading zeros (e.g., ``'0001'``,
        ``'0002'``) rather than as an integer to preserve the COBOL
        label semantics and maintain parity with
        ``DisclosureGroup.tran_cat_cd`` and
        ``TransactionCategory.cat_cd`` for clean composite-key
        joins.
    balance : Decimal
        Running balance for this (account, type, category) tuple
        (from COBOL ``TRAN-CAT-BAL``, ``PIC S9(09)V99`` — 9 integer
        digits + 2 fractional digits, signed). Stored as PostgreSQL
        ``NUMERIC(15, 2)`` and returned as a ``decimal.Decimal``
        value — **never** as a Python ``float`` — so that the
        canonical interest formula ``(balance × int_rate) / 1200``
        from ``CBACT04C.cbl`` preserves exact decimal precision.
        Defaults to ``Decimal("0.00")`` for safe arithmetic-neutral
        semantics when a brand-new ``(acct_id, type_cd, cat_cd)``
        combination is first written by the POSTTRAN job.
    """

    __tablename__ = "transaction_category_balance"

    # ------------------------------------------------------------------
    # Primary key part 1 of 3: account identifier
    # (COBOL ``TRANCAT-ACCT-ID`` PIC 9(11))
    #
    # Stored as String(11) — NOT as an integer — to preserve leading-
    # zero formatting (e.g., '00000000001') consistent with
    # ``Account.acct_id``. This keeps the column semantically a
    # fixed-width label rather than a numeric quantity and avoids
    # lossy round-trips when joining against the ``account`` table.
    # ------------------------------------------------------------------
    acct_id: Mapped[str] = mapped_column(
        String(11),
        primary_key=True,
    )

    # ------------------------------------------------------------------
    # Primary key part 2 of 3: transaction-type code
    # (COBOL ``TRANCAT-TYPE-CD`` PIC X(02))
    #
    # Logical FK to ``transaction_type.tran_type``. Typical values:
    #   '01' Purchase · '02' Payment · '03' Credit · '04' Debit
    #   '05' Refund   · '06' Adjustment · '07' Fee
    # ------------------------------------------------------------------
    type_cd: Mapped[str] = mapped_column(
        String(2),
        primary_key=True,
    )

    # ------------------------------------------------------------------
    # Primary key part 3 of 3: transaction-category code
    # (COBOL ``TRANCAT-CD`` PIC 9(04))
    #
    # Stored as String(4) with leading zeros (e.g. '0001') rather than
    # as an integer — preserves the COBOL label semantics and matches
    # the type used by ``TransactionCategory.cat_cd`` and
    # ``DisclosureGroup.tran_cat_cd`` for clean composite-key joins.
    # Together with ``type_cd`` this forms a logical FK to
    # ``transaction_category.(type_cd, cat_cd)``.
    # ------------------------------------------------------------------
    cat_cd: Mapped[str] = mapped_column(
        String(4),
        primary_key=True,
    )

    # ------------------------------------------------------------------
    # Running balance (COBOL ``TRAN-CAT-BAL`` PIC S9(09)V99)
    #
    # COBOL picture breakdown:
    #   S       — signed
    #   9(09)   — 9 integer digits
    #   V       — implicit decimal point
    #   99      — 2 fractional digits
    # Total digits: 9 + 2 = 11  →  PostgreSQL NUMERIC(15, 2)
    #   (15 digits gives generous headroom above the COBOL 11-digit
    #   maximum for any downstream widening of the monetary range.)
    #
    # FINANCIAL-PRECISION CRITICAL (AAP §0.7.2):
    #   * SQLAlchemy returns this column as ``decimal.Decimal``.
    #   * Do NOT cast to ``float`` at any point in the posting,
    #     interest-calculation, or statement-generation pipelines —
    #     floating-point arithmetic is forbidden for financial
    #     calculations.
    #   * Default ``Decimal("0.00")`` ensures a new
    #     ``(acct_id, type_cd, cat_cd)`` row — created when POSTTRAN
    #     encounters a never-before-seen category combination for
    #     an account — is arithmetic-neutral and produces correct
    #     posting/interest results from the first transaction on.
    # ------------------------------------------------------------------
    balance: Mapped[Decimal] = mapped_column(
        Numeric(15, 2),
        nullable=False,
        default=Decimal("0.00"),
    )

    # Note: COBOL ``FILLER PIC X(22)`` — the trailing 22 bytes of
    # padding in the original 50-byte VSAM record — is deliberately NOT
    # mapped. In the relational model, column widths are explicit and
    # trailing padding has no storage or semantic meaning.

    def __repr__(self) -> str:  # pragma: no cover - trivial formatting
        """Return a developer-friendly string representation.

        Includes all four mapped columns (the 3-part composite
        primary key and the running balance). The ``balance`` is
        formatted via ``repr()`` which preserves full ``Decimal``
        precision — no rounding, no scientific notation, and no
        implicit conversion to ``float``.

        Returns
        -------
        str
            Representation of the form
            ``TransactionCategoryBalance(acct_id='00000000001',
            type_cd='01', cat_cd='0001',
            balance=Decimal('1234.56'))``.
        """
        return (
            f"TransactionCategoryBalance("
            f"acct_id={self.acct_id!r}, "
            f"type_cd={self.type_cd!r}, "
            f"cat_cd={self.cat_cd!r}, "
            f"balance={self.balance!r}"
            f")"
        )

File: src/main/java/com/aws/carddemo/entity/TransactionCategoryBalance.java
/*
 * Copyright Amazon.com, Inc. or its affiliates.
 * All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 */
package com.aws.carddemo.entity;

import jakarta.persistence.Column;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.util.Objects;

/**
 * JPA entity that replaces the COBOL {@code TCATBAL} VSAM KSDS file's
 * {@code TRAN-CAT-BAL-RECORD} record described by
 * {@code app/cpy/CVTRA01Y.cpy} (RECLN 50). Seed data is sourced from
 * {@code app/data/ASCII/tcatbal.txt} (50 rows × 50 bytes = 2500 bytes).
 *
 * <h2>COBOL Provenance — CVTRA01Y.cpy</h2>
 *
 * <p>The original copybook layout is a fixed-width 50-byte record:
 * <pre>
 *   01 TRAN-CAT-BAL-RECORD.
 *      05 TRAN-CAT-KEY.
 *         10 TRANCAT-ACCT-ID    PIC 9(11).       --&gt; {@link TransactionCategoryBalanceKey#getTrancatAcctId()}
 *         10 TRANCAT-TYPE-CD    PIC X(02).       --&gt; {@link TransactionCategoryBalanceKey#getTrancatTypeCd()}
 *         10 TRANCAT-CD         PIC 9(04).       --&gt; {@link TransactionCategoryBalanceKey#getTrancatCd()}
 *      05 TRAN-CAT-BAL          PIC S9(09)V99.   --&gt; {@link #tranCatBal} (BigDecimal scale 2)
 *      05 FILLER                PIC X(22).       --&gt; (no Java mapping — padding only)
 * </pre>
 *
 * <h2>CBACT04C — Interest Calculation Update Path</h2>
 *
 * <p>The 50 canonical rows in {@code app/data/ASCII/tcatbal.txt} are
 * read, updated, and rewritten by the COBOL {@code CBACT04C} interest
 * calculator (migrated to
 * {@code com.aws.carddemo.batch.InterestCalculationProcessor} per AAP
 * §0.5.1). For each row the COBOL program:
 * <ol>
 *   <li>READs the {@code TCATBAL} record by composite key.</li>
 *   <li>Looks up the matching {@code DIS-INT-RATE} in the {@code DISCGRP}
 *       catalog (the {@link DiscountGroup} JPA entity) using the
 *       account's {@code ACCT-GROUP-ID} plus the {@code TRANCAT-TYPE-CD}
 *       and {@code TRANCAT-CD} from this row's composite key.</li>
 *   <li>Computes {@code MONTHLY-INTEREST = TRAN-CAT-BAL * DIS-INT-RATE /
 *       1200} with {@code RoundingMode.HALF_EVEN} preserving scale 2
 *       (AAP §0.10.3 financial-precision contract).</li>
 *   <li>Writes a new {@link Transaction} record carrying the computed
 *       interest amount.</li>
 *   <li>REWRITEs the {@code TCATBAL} record with {@code TRAN-CAT-BAL}
 *       set back to {@code 0} (since the interest amount is now
 *       captured in the transaction journal and the category balance
 *       restarts for the next accrual cycle).</li>
 * </ol>
 *
 * <p>Skip conditions:
 * <ul>
 *   <li>{@code ACCT-GROUP-ID = "ZEROAPR   "} — the
 *       {@link DiscountGroup#getDisIntRate()} is {@code 0.00} so the
 *       computed interest is zero and the REWRITE is a no-op.</li>
 *   <li>{@code DISCGRP} lookup returns {@code DFHRESP(NOTFND)} — the
 *       processor re-issues the lookup with {@code "DEFAULT   "} as the
 *       account group, falling back to the catch-all rate.</li>
 * </ul>
 *
 * <h2>Composite Key — Embedded ID</h2>
 *
 * <p>The {@link #key} field carries the composite primary key as a
 * separate {@link TransactionCategoryBalanceKey} value object (the
 * {@code @EmbeddedId} pattern). The composite-key class is the natural
 * Java mapping of the COBOL {@code TRAN-CAT-KEY} group-level item, and
 * keeping it as a discrete value object (rather than three scalar fields
 * on this entity) makes the composite identity surface explicit at the
 * type-system level: methods that operate on a transaction-category
 * balance row accept a single {@link TransactionCategoryBalanceKey}
 * argument rather than a triplet of scalars where call-sites could
 * accidentally swap arguments.
 *
 * <h2>Financial Precision (AAP §0.10.3)</h2>
 *
 * <p>The {@link #tranCatBal} field carries the {@code TRAN-CAT-BAL
 * PIC S9(09)V99} signed packed-decimal value. The COBOL field is signed
 * with 9 integer digits + 2 fractional digits (theoretical range
 * {@code -999_999_999.99} to {@code +999_999_999.99}), mapped to a Java
 * {@link BigDecimal} with scale 2 per AAP §0.10.3 ("No float or double
 * used for any monetary value — BigDecimal exclusively"). The
 * REFACTOR-flavor migration agent will add the
 * {@code @Column(name = "tran_cat_bal", precision = 11, scale = 2,
 * nullable = false)} JPA annotation to enforce the scale at the JDBC
 * boundary; the Flyway DDL will create the underlying PostgreSQL column
 * as {@code NUMERIC(11, 2)} (9 integer digits + 2 fractional digits) to
 * mirror the COBOL precision contract.
 *
 * <h2>Design Note — Stub Status</h2>
 *
 * <p>This class is a <strong>minimum-viable POJO</strong> created to
 * satisfy {@link com.aws.carddemo.repository.TransactionCategoryBalanceRepository}
 * compilation and the {@code TransactionCategoryBalanceRepositoryIT}
 * integration test suite per AAP §0.5.1. Subsequent migration agents
 * (REFACTOR flavor) will add JPA annotations ({@code @Entity},
 * {@code @EmbeddedId}, {@code @Column},
 * {@code @Table(name = "transaction_category_balances")}) and Bean
 * Validation constraints ({@code @DecimalMin}/{@code @DecimalMax} on
 * {@link #tranCatBal} to enforce the {@code PIC S9(09)V99} numeric
 * range) once the entity is wired into the Hibernate
 * {@code SessionFactory}. The Flyway scripts under
 * {@code src/main/resources/db/migration/} (also REFACTOR-flavor) will
 * create the {@code transaction_category_balances} table (with the
 * composite primary key {@code (trancat_acct_id CHAR(11),
 * trancat_type_cd CHAR(2), trancat_cd INTEGER)} and the balance column
 * {@code tran_cat_bal NUMERIC(11, 2) NOT NULL}) and populate it with
 * the 50 reference rows from {@code app/data/ASCII/tcatbal.txt}.
 *
 * <h2>Security — toString() Redaction (AAP §0.10.5)</h2>
 *
 * <p>{@link #toString()} surfaces the composite key (via
 * {@link TransactionCategoryBalanceKey#toString()}) but
 * <strong>does NOT include</strong> the monetary {@link #tranCatBal}
 * field. AAP §0.10.5 ("No financial data written to logs at any level")
 * mandates that balance values must not appear in any diagnostic
 * {@code toString()} output that could be captured by a logger or
 * exception stack trace. The redaction is enforced by deliberate
 * omission: callers that need to surface the balance value do so
 * through explicit getter calls in audited code paths, never through
 * the entity's diagnostic representation.
 *
 * @see TransactionCategoryBalanceKey
 * @see com.aws.carddemo.repository.TransactionCategoryBalanceRepository
 */
@Entity
@Table(name = "transaction_category_balances")
public class TransactionCategoryBalance {

    /**
     * Composite primary key holding the
     * {@code (TRANCAT-ACCT-ID, TRANCAT-TYPE-CD, TRANCAT-CD)} 3-tuple.
     * The {@link EmbeddedId} annotation tells Hibernate to materialise
     * the value object from the three key columns of the
     * {@code transaction_category_balances} table on every read and to
     * decompose it back into the three columns on every write.
     */
    @EmbeddedId
    private TransactionCategoryBalanceKey key;

    /**
     * The {@code TRAN-CAT-BAL} balance value per {@code CVTRA01Y.cpy}
     * ({@code PIC S9(09)V99}). Carries the running per-category balance
     * for the parent account, updated by the {@code CBACT04C} interest
     * calculator (migrated to
     * {@code com.aws.carddemo.batch.InterestCalculationProcessor}).
     *
     * <p>The {@link BigDecimal} type and the scale-2 contract are
     * mandated by AAP §0.10.3 ("No float or double used for any monetary
     * value — BigDecimal exclusively"; "BigDecimal rounding mode set to
     * HALF_EVEN (banker's rounding) matching COBOL PICTURE clause
     * precision"). The COBOL {@code PIC S9(09)V99} field allows up to 9
     * integer digits + 2 fractional digits (theoretical range
     * {@code -999_999_999.99} to {@code +999_999_999.99}); the Flyway
     * column type {@code NUMERIC(11, 2)} preserves both the precision
     * and the scale at the storage boundary.
     */
    @Column(name = "tran_cat_bal", precision = 11, scale = 2, nullable = false)
    private BigDecimal tranCatBal;

    /**
     * Default no-arg constructor (required by JPA reflection-based
     * instantiation when Hibernate materialises rows from a query result
     * set).
     */
    public TransactionCategoryBalance() {
        // intentionally empty
    }

    /**
     * @return the composite primary key carrying the
     *         {@code (TRANCAT-ACCT-ID, TRANCAT-TYPE-CD, TRANCAT-CD)}
     *         3-tuple
     */
    public TransactionCategoryBalanceKey getKey() {
        return key;
    }

    /**
     * @param key the composite primary key carrying the
     *            {@code (TRANCAT-ACCT-ID, TRANCAT-TYPE-CD, TRANCAT-CD)}
     *            3-tuple
     */
    public void setKey(TransactionCategoryBalanceKey key) {
        this.key = key;
    }

    /**
     * @return the {@code TRAN-CAT-BAL} balance value with scale 2
     *         (BigDecimal — AAP §0.10.3 financial-precision contract)
     */
    public BigDecimal getTranCatBal() {
        return tranCatBal;
    }

    /**
     * @param tranCatBal the {@code TRAN-CAT-BAL} balance value with
     *                   scale 2 (BigDecimal — AAP §0.10.3
     *                   financial-precision contract)
     */
    public void setTranCatBal(BigDecimal tranCatBal) {
        this.tranCatBal = tranCatBal;
    }

    /**
     * Equality is based on the composite primary key {@link #key} alone.
     * JPA-managed entities are considered equal iff they share the same
     * primary key value; the balance is deliberately excluded from
     * equality so transient and managed copies of the same logical
     * reference row compare equal regardless of whether one carries a
     * pre-update balance and the other a post-update balance (the row
     * is the same row, identified by the composite key).
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof TransactionCategoryBalance)) {
            return false;
        }
        TransactionCategoryBalance other = (TransactionCategoryBalance) o;
        return Objects.equals(key, other.key);
    }

    /**
     * Hash by composite primary key, consistent with {@link #equals(Object)}.
     */
    @Override
    public int hashCode() {
        return Objects.hash(key);
    }

    /**
     * Diagnostic string surfacing the composite key only.
     *
     * <p><strong>Security (AAP §0.10.5).</strong> The monetary
     * {@link #tranCatBal} field is deliberately omitted from this output
     * because AAP §0.10.5 ("No financial data written to logs at any
     * level") forbids balance values from appearing in any diagnostic
     * representation that could leak into a logger or stack trace.
     * Callers that need to surface the balance do so through explicit
     * getter calls in audited code paths, never through this entity's
     * diagnostic representation.
     */
    @Override
    public String toString() {
        return "TransactionCategoryBalance{"
                + "key=" + key
                + ", tranCatBal=[REDACTED — AAP §0.10.5 no financial data in logs]"
                + '}';
    }
}
